File: 2024/3/solution.js
const fs = require("fs");
const path = require("path");

const input = fs
  .readFileSync(path.join(__dirname, "input.txt"), "utf-8")
  .split("\n");

const regex = /mul\((\d{1,3}),(\d{1,3})\)|do(?:n\'t)?\(\)/g;

let part1 = 0;
let part2 = 0;

let doMul = true;
for (const line of input) {
  let match;
  while ((match = regex.exec(line))) {
    if (match[0] === "do()") {
      doMul = true;
    } else if (match[0] === "don't()") {
      doMul = false;
    } else {
      const prod = Number(match[1]) * Number(match[2]);
      part1 += prod;
      part2 += doMul ? prod : 0;
    }
  }
}

console.log(part1, part2);

File: 2024/3/solution.go
package main

import (
	"fmt"
	"os"
	"path/filepath"
	"regexp"
	"runtime"
	"strconv"
)

func main() {
	_, filename, _, _ := runtime.Caller(0)
	dirname := filepath.Dir(filename)
	inputFilePath := filepath.Join(dirname, "input.txt")
	data, err := os.ReadFile(inputFilePath)
	if err != nil {
		panic(err)
	}
	puzzleInput := string(data)
	fmt.Println(part1(puzzleInput))
	fmt.Println(part2(puzzleInput))
}

func part1(puzzleInput string) int {
	res := 0
	mulRegex := regexp.MustCompile(`mul\((\d+),(\d+)\)`)
	matches := mulRegex.FindAllStringSubmatch(puzzleInput, -1)
	for _, match := range matches {
		var n1, n2 int
		var err error
		if n1, err = strconv.Atoi(match[1]); err != nil {
			panic(err)
		}
		if n2, err = strconv.Atoi(match[2]); err != nil {
			panic(err)
		}
		res += n1 * n2
	}

	return res
}

func part2(puzzleInput string) int {
	removeRegex := regexp.MustCompile(`(?s)don\'t\(\).*?(?:do\(\)|$)`)
	puzzleInput2 := removeRegex.ReplaceAllString(puzzleInput, "")
	return part1(puzzleInput2)
}
